### PR TITLE
Phantom let support in [Selectgen]

### PR DESCRIPTION
**Pull request by mshinwell**
This patch adds support for transferring phantom let information from Cmm expressions to Mach instructions.  This is done by keeping track of which phantom identifiers are currently in scope and tagging such information onto the generated instructions.  The defining expressions corresponding to the phantom identifiers are kept separately in the `fundecl` record.

@lthls Maybe you could look at this one?

### CODE REVIEW

File: asmcomp/selectgen.ml [hunk 1]
```diff
@@ -1437,4 +1549,6 @@ let _ =
   Simplif.is_tail_native_heuristic := is_tail_call
 
 let reset () =
-  current_function_name := ""
+  current_function_name := "";
+  Ident.Tbl.clear phantom_lets;
+  dead_phantom_lets := Ident.Set.empty
```
Comment (lthls) [inline asmcomp/selectgen.ml:1549]: As a safe guard, I would also reset the global tables in `emit_fundecl` above. I'm not convinced `reset` is systematically called between two function declarations.

**Review by lthls — APPROVED** [review]
I have reviewed this PR and believe it is correct (save for a small worry) and matches the description.

Note that this PR includes commits submitted as other PRs, but some of those PRs have been updated since, so this one shouldn't be merged until it has been rebased on top of the (hopefully merged) others.

File: asmcomp/selectgen.ml (hunk 1 above)
Comment (mshinwell) [reply asmcomp/selectgen.ml:1549]: I've changed the code as you suggest.

**Comment by mshinwell** [comment]
This has now been rebased.

**Comment by mshinwell** [comment]
(Rebased again)

**Comment by chambart** [comment]
I didn't review it myself but I trust @lthls on this. Good to go as soon as the conflict on Changes is fixed.

File: asmcomp/selectgen.ml [hunk 2]
```diff
@@ -815,21 +971,37 @@ method emit_expr (env:environment) exp =
...
-        let (r, s) = self#emit_sequence new_env e2 in
         (nfail, (r, s))
       in
       let l = List.map translate_one_handler handlers in
       let a = Array.of_list ((r_body, s_body) :: List.map snd l) in
-      let r = join_array a in
+      let r = join_array env a ~bound_name in
       let aux (nfail, (_r, s)) = (nfail, s#extract) in
-      self#insert (Icatch (rec_flag, List.map aux l, s_body#extract)) [||] [||];
+      self#insert env (Icatch (rec_flag, List.map aux l, s_body#extract)) [||] [||];
```
Comment (damiendoligez) [inline asmcomp/selectgen.ml:1004]: This line is over 80 chars.

File: asmcomp/selectgen.ml [hunk 3]
```diff
@@ -1145,52 +1352,135 @@ method emit_tail (env:environment) exp =
...
+                    is_assignment = false;
+                  }
+                in
+                seq#insert_debug env (Iop naming_op) Debuginfo.none r [| |])
+              ids_and_rs)
+        in
+        nfail, seq
       in
-      self#insert (Icatch(rec_flag, List.map aux handlers, s_body)) [||] [||]
+      self#insert env (Icatch(rec_flag, List.map aux handlers, s_body)) [||] [||]
```
Comment (damiendoligez) [inline asmcomp/selectgen.ml:1378]: This line is over 80 chars.

**Review by damiendoligez — APPROVED** [review]
Approved based on @lthls's review. Cut the two overlong lines, fix the conflict in Changes, and feel free to merge.

**Comment by mshinwell** [comment]
@lthls I'm afraid I think I am going to have to make a new version of this.  However the comments above have been / will be incorporated.

**Pull request closed without merging**